### ISSUE TITLE
form emitting twice on save

### ISSUE THREAD

**Issue by thamera** [MrWolfZ/ngrx-forms#226]
I have a form that is emitting twice on the save (submit) event I have setup.  I have submits calling a save function in my controller in the same was as the examples.  I am trying to simply get the current value from the store, post it to an external rest API (promise wrapped in observable) and then dispatch an action to unblock the form (using ngrx-forms recursive sample)  What I am seeing is the console I inserted is responding twice and the unblocking dispatch then of course shows twice in the store actions.

```
save(): void {
    this.store.dispatch(new BlockInputsAction());

    this.formState$.pipe(
      take(1),
      filter(s => s.isValid),
      map(fs => {

        console.dir(fs);
        return from(t.set(TrelloScope.card, TrelloVisibility.shared, fs.value)
        .then(() => this.store.dispatch(new UnBlockInputsAction()))
        );
      }),
    ).subscribe();
```

![image](https://user-images.githubusercontent.com/5296746/104955209-b3aad200-597e-11eb-9541-019d9e7f19a8.png)

bear with me as I am still trying to get my arms wrapped around both NGRX and Ngrx-Forms.  Just wanted to see if anyone had an initial though before I dive deeper.

**Comment by MrWolfZ** [comment]
Apologies for the late answer. It is indeed strange that the code you posted would log the form state twice. What's confusing to me as well is that I don't see the `BlockInputs` action in the history of actions. Are you sure the `save` function is properly called and the logging and unblocking aren't happening somewhere else?

Also, how is `this.formState$` defined? Observables are a bit weird in that you can end up with duplicate subscriptions with certain setups.

In any case, I think you can tweak your code a bit, and maybe (even though unlikely) it will help with the duplicate event. The `map` returning an `Observable` is not correct, you want to use `mergeMap` instead. You can also flatten some of the other code.

```ts
save(): void {
    this.store.dispatch(new BlockInputsAction());

    this.formState$.pipe(
      take(1),
      filter(s => s.isValid),
      tap(fs => console.log(fs)),
      mergeMap(fs => from(t.set(TrelloScope.card, TrelloVisibility.shared, fs.value))
      map(() => new UnBlockInputsAction()),
    ).subscribe(this.store);
```

**Comment by thamera** [comment]
My mistake on the block inputs action.  I had disabled it while troubleshooting and sent that screenshot so it was on purpose.  It does fire when turned on!

save function is getting called. the formstate is defined. `formState$: Observable<FormGroupState<FormValue>>;`
and then in the constructor:  `this.formState$ = store.pipe(select(s => s.milestone.formState));`

in my ngoninit, after getting data, i prepopulte the form with:
`const milestone: FormValue = {
          milestone_start: data.milestone_start,
          milestone_anticipated: data.milestone_anticipated,
          milestone_actual: data.milestone_actual,
          milestone_status: data.milestone_status,
          milestone_parent: data.milestone_parent,
          milestone_predecessor: data.milestone_predecessor,
        };
        this.store.dispatch(new SetInitialValueAction(
          milestone
        ));
`

I did apply the suggested code changes and still getting the duplication.  If it helps, here is some of my reducer:

`const reducers = combineReducers<State['milestone'], any>({
  formState(state = initialState, a: Action): FormGroupState<FormValue> {
    const form = validationFormGroupReducer(state, a);

    switch (a.type) {
      case `${FORM_ID} SET_INITIAL_VALUE`:
        console.log('set initial value...');
        const payload = (a as SetInitialValueAction).initialValue;
        state = updateRecursive(
            enable(state),
            s => s.userDefinedProperties.wasDisabled ? disable(s) : s,
        );
        const updatedForm = setValue({
          ...state.value,
          ...payload
        })(state);
        return updatedForm;

      case `${FORM_ID} BLOCK_INPUTS`:
        console.log('block inputs...');
        state = updateRecursive(
          state,
          s => setUserDefinedProperty(s, 'wasDisabled', s.isDisabled),
        );
        return disable(state);

      case `${FORM_ID} UN_BLOCK_INPUTS`:
        console.log('unlock inputs...');
        state = enable(state);
        return updateRecursive(
          state,
          s => s.userDefinedProperties.wasDisabled ? disable(s) : s,
        );

      default: {
        return form;
      }
    }
  },
  submittedValue(state: FormValue | undefined, a: SetSubmittedValueAction): FormValue | undefined {
    switch (a.type) {
      case SetSubmittedValueAction.TYPE:
        return a.submittedValue;
      default:
        return state;
    }
  },
});`

**Comment by MrWolfZ** [comment]
Your code looks good to me, so without a reproduction example it is difficult to help. Can you please try to create a repository where this problem can be reproduced? In the best case, you could just share your app (if possible) and strip everything except the problematic code from it. Alternatively, you can fork [this code sandbox](https://codesandbox.io/s/ngrx-forms-issue-template-bw7n4) and reproduce the issue there.

**Comment by thamera** [comment]
Took a bit to setup as I wanted to mock my service and globals and include the root store to make it as similar to the actual app as possible.  You should be able to see it at the following.  Console in code sandbox shows the duplication.

https://codesandbox.io/s/ngrx-forms-issue-template-forked-tqzm4

**Comment by thamera** [comment]
Nevermind!  Found it (and now smacking myself).  I tend to avoid form submission and just run button actions to handle.  Since I copied code, I didn't realize my submit button has a click event and was submit type so the save action was getting hit twice.  I will say going through this makes me much more comfortable with ngrx-forms and I'm sold (I already have half a dozen forms in place).  Thanks for all the hard work putting this together and for help with my issue (even though it was self inflicted!)
